src/Evolution/Systems/CurvedScalarWave/Equations.hpp

```diff
@@ -93,4 +87,203 @@ struct ComputeDuDt {
+      const Scalar<DataVector>& pi, const tnsr::i<DataVector, Dim>& phi,
+      const Scalar<DataVector>& psi, const Scalar<DataVector>& gamma1,
+      const Scalar<DataVector>& gamma2, const Scalar<DataVector>& lapse,
+      const tnsr::I<DataVector, Dim>& shift,
+      const tnsr::II<DataVector, Dim>& inverse_spatial_metric,
+      const tnsr::i<DataVector, Dim>& interface_unit_normal) noexcept;
+};
+
+/*!
+ * \ingroup NumericalFluxesGroup
+ * \brief Computes the upwind flux for scalar-waves in curved spacetime.
+ *
+ * \details The upwind flux is given by \cite Teukolsky2015ega :
+ *
+ * \f[
+ * G = S\left(\Lambda^+ S^{-1} U^{\rm int}
+ *           + \Lambda^- S^{-1} U^{\rm ext}\right)
+ * = S \left(\Lambda^+ \hat{U}^{\rm int}
+ *           + \Lambda^- \hat{U}^{\rm ext}\right),
+ * \f]
+ *
+ * where
+ *
+ *  - \f$G\f$ is the numerical upwind flux dotted with the interface normal;
+ *  - \f$U\f$ is a vector of all evolved variables;
+ *  - \f$S\f$ is a matrix whose columns are the eigenvectors of the
+ *       characteristic matrix for the evolution system. It maps the
+ *       evolved variables to characteristic variables \f$\hat{U}\f$, s.t.
+ *       \f$\hat{U} := S^{-1}\cdot U\f$; and
+ *  - \f$\Lambda^\pm\f$ are diagonal matrices containing
+ *       positive / negative eigenvalues of the same matrix as its elements.
+ *
+ * The superscripts \f${\rm int}\f$ and \f${\rm ext}\f$ on \f$U\f$ indicate
+ * that the corresponding set of variables at the element interface have been
+ * taken from the _interior_ or _exterior_ of the element. Exterior of the
+ * element is naturally the interior of its neighboring element. Therefore,
+ * \f$\hat{U}^{\rm int}:= S^{-1}U^{\rm int}\f$ are the characteristic variables
+ * at the element interface computed using evolved variables from the interior
+ * of the element, and \f$\hat{U}^{\rm ext}= S^{-1}U^{\rm ext}\f$ are the
+ * same computed from evolved variables taken from the element exterior, i.e.
+ * the neighboring element.
+ * Further, the sign of characteristic speeds indicates the direction of
+ * propagation of the corresponding characteristic field with respect to the
+ * interface normal that the field has been
+ * computed along (with negative speeds indicating incoming characteristics,
+ * and vice-versa). Therefore, \f$\Lambda^+\f$ contains characterstic speeds
+ * for variables that are outgoing from the element at the interface in
+ * question, and vice-versa for \f$\Lambda^-\f$. An ambiguity naturally arises
+ * as to which set of
+ * evolved variables (''interior'' or ''exterior'') to use when computing these
+ * speeds. Following the treatment of generalized harmonic evolution system's
+ * characteristics in its upwind flux (\ref GeneralizedHarmonic::UpwindFlux ),
+ * here also we compute both and use their average, i.e.
+ * \f$\lambda^{\rm avg} = (\lambda^{\rm int} + \lambda^{\rm ext})/2\f$, to
+ * populate \f$\Lambda^\pm\f$.
+ *
```

Comment:
I think this treatment is fine as long as there are no shocks, as is the case for GH. Is that the case for scalar wave on a curved background?